### ISSUE TITLE
Deploying to Amazon Elastic Beanstalk

### ISSUE THREAD

**Issue by andybeak** [laradock/laradock#244]
Has anybody had any experience with this?  Could you recommend some reading sources or point me in the direction of example config setups?

**Comment by Spracto** [comment]
I'm currently struggling through trying to do this. I'd appreciate some help with it as I'm totally lost trying to write a Dockerrun.aws.json file for the 5 containers that laradock created to dockerize the laravel project.

**Comment by jamesalton** [comment]
Spracto: I asked Amazon what the best way to do this and they told me (in a nutshell): ELB is crap, use ECS instead. (Even though AWS EB seems to have some sort of provision for Multi-container docker, apparently it doesn't support compose like ECS does.) So, I'm struggling through `ecs-cli compose` now. Not easy. I found a script called "build-tag-push" which seems to put the docker-compose.yml into a format of images only without builds (apparently ECS can't support that!) The problem I am facing now is: apparently ECS can't support the `tianon/true` "data-only container". It feels a bit like a whack-a-mole of problems and I'm fairly surprised nobody else has solved this yet. Amazon's solutions are all feeling broken to me after going through all of this. I guess I am at "Here be dragons", even though I'm just trying to do something that feels standard like Docker.

**Comment by ivanhueso** [comment]
I'm trying to deloy to Elastic Beanstalk I'm having issues when it too. Need help.

**Comment by davehensley** [comment]
1. Create whatever images you need, e.g. `docker-compose build nginx php-fpm`
1. Tag your image, e.g. `docker tag laradock_nginx:latest 123456789.dkr.ecr.us-east-1.amazonaws.com/laradock/nginx:latest` (replace with the URI of your ECR repository)
1. Repeat the previous step for the rest of your images.
1. Get your authorization token from AWS: `aws ecr get-login`
1. Run the docker command that the aws tool outputs in the previous step: `docker login -u AWS -p ...`
1. Deploy your new image to ECR: `docker push 123456789.dkr.ecr.us-east-1.amazonaws.com/laradock/nginx:latest` (replace with the URI of your repository)
1. Repeat the previous step for the rest of your images.
1. Create a file called `Dockerrun.aws.json` one directory up from your Laravel app (so if Laravel is in `~/Projects/YourProject/application/` then put `Dockerrun.aws.json` in `~/Projects/YourProject/`. Use the sample file below as a starting point.
1. Deploy your app: `eb deploy`

Sample `Dockerrun.aws.json` file:
```
{
  "AWSEBDockerrunVersion": 2,
  "containerDefinitions": [
    {
      "essential": true,
      "image": "0123456789.dkr.ecr.us-east-1.amazonaws.com/laradock/php-fpm:latest",
      "memory": 256,
      "mountPoints": [
        {
          "containerPath": "/var/www",
          "sourceVolume": "application"
        },
        {
          "containerPath": "/usr/local/var/log",
          "sourceVolume": "awseb-logs-php-fpm"
        }
      ],
      "name": "php-fpm"
    },
    {
      "essential": true,
      "image": "0123456789.dkr.ecr.us-east-1.amazonaws.com/laradock/nginx:latest",
      "links": [
        "php-fpm"
      ],
      "memory": 512,
      "mountPoints": [
        {
          "containerPath": "/var/www",
          "sourceVolume": "application"
        },
        {
          "containerPath": "/var/log/nginx",
          "sourceVolume": "awseb-logs-nginx"
        }
      ],
      "name": "nginx",
      "portMappings": [
        {
          "containerPort": 80,
          "hostPort": 80
        }
      ]
    }
  ],
  "volumes": [
    {
      "host": {
        "sourcePath": "/var/app/current/application"
      },
      "name": "application"
    }
  ]
}
```

**Comment by Mahmoudz** [comment]
@davehensley can you include your answer in the docs!?

**Comment by davehensley** [comment]
> @davehensley can you include your answer in the docs!?

@Mahmoudz I'm not sure which docs you're referring to ([laradock.io](http://laradock.io/) maybe?), but I'm not part of the Laradock project so I don't have write access to them. But if you do, you're more than welcome to copy my solution and include it there if you think it would be useful.

**Comment by Mahmoudz** [comment]
@davehensley actually the site is hosted on Github and the readme files are in the project, so you can send a normal PR as you used to when updating the readme.md. Please refer to this docs for how to update the docs [http://laradock.io/contributing/#update-documentation-site](http://laradock.io/contributing/#update-documentation-site). Note: do not edit HTML files, someone did that mistake before, just edit the markdown files and re-generate the HTML from them, the link should help. Thanks for your quick response. Best,

**Comment by canfone** [comment]
@davehensley Thanks for your answer. In that case, do I need to create each repository for each images in my ECR? For example, one for nginx, one for php-fpm, etc? Or I can create just one repository and tag all of them in that one repo? 

Sorry for my ignorance. Hope you can help me.

**Comment by davehensley** [comment]
@canfone You should create a separate repo for each image.

**Comment by canfone** [comment]
@davehensley 
Thanks for your quick response.
I followed your instructions but I encounter this error when deploying to eb
```
Dockerfile and Dockerrun.aws.json are both missing, abort deployment
```

I believe that there are processes I need to do before the steps you suggested. I don't know where or what to start. Do you have any suggestions or ideas what things I need to be done before following the steps you gave?

Thanks.

**Comment by davehensley** [comment]
@canfone Hmm,  what was the process that you used to create your EB app? Try something like this (it should download Laravel and deploy the default app):

```
~ $ mkdir -p ~Projects/my-project && cd !$
my-project $ eb init
my-project $ git clone git@github.com:laravel/laravel.git application
my-project $ cp /path/to/Dockerrun.aws.json .
my-project $ eb create
```

**Comment by canfone** [comment]
@davehensley Thanks for your reply.

I'm using awsebcli to create my EB app. I think I managed to create the app and deployed it to elastic beanstalk. However, my nginx configurations were not included in the nginx image of my ECR. Also, the application files is not in `/var/www` inside workspace container. It is in `/var/app/current` instead. Maybe there are bindings I missed that's why those files were not properly located in my containers.

**Comment by canfone** [comment]
@davehensley Nevermind my previous comment. I managed to mount nginx conf files. I just need to use volumes. Thanks for your help.

**Comment by chkm8** [comment]
Hi, Im follwing the guide that is provided above, i managed to create the laradock instance in elastic server via ecr. but the application is not working, im having 502 bad gateway error, and when I checked the apache2 container logs inside eb server, is says that SSLCertificateFile: file '/etc/ssl/certs/ssl-cert-snakeoil.pem' does not exist or is empty, how would i solve this one, in my local instance there is no ssl-cert issues, so the logic would be my local should be the same once i upload it on eb server. hope you could help me . thanks

**Comment by davehensley** [comment]
@chkm8 This is a bit outside the scope of the original question, but the answer is that it's easiest to set up a secure site on Elastic Beanstalk if the EC2 instances inside your EB environment don't know anything about SSL. Even if you're running a secure site, your webserver inside your docker container should be listening on port 80 only, so you don't even need to install mod_ssl. The job of storing SSL certificates and keys and encrypting the connection is performed by Amazon's Elastic Load Balancer.

So the traffic goes like this:

Your container  <---port 80 (http)---> Elastic Load Balancer <---port 443 (https)---> Internet

AFAIK there isn't a way to upload an SSL certificate/key through the AWS console, but you can do it pretty easily with the aws command line tool:
http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/configuring-https-ssl-upload.html

Everything else is configurable through the AWS console (just allow incoming https connections in your Elastic Load Balancer's security group, and then enable HTTPS on port 443 in your load balancer settings and choose the certificate that you just uploaded from the dropdown menu).

If you really do need full end-to-end encryption (which you almost certainly don't), or you're running a single-instance EB environment without a load balancer, then it is possible to configure your container to listen for https connections (but I strongly recommend trying the other way first):
http://docs.aws.amazon.com/elasticbeanstalk/latest/dg/configuring-https-endtoend.html

**Comment by chkm8** [comment]
@davehensley , yup i finally uploaded our project in EB following the guide you provided, the error I was referring to is how the "laradock apache2 container" was buid making a copy of the default ssl and with missing plugins (snakeoil). I rebuild a new one but remove the default ssl.

**Comment by hareku** [comment]
Hi. I tried the deployment to elastic beanstalk.
(Nginx + PHP-FPM + PHP-worker)

```
{
  "AWSEBDockerrunVersion": 2,
  "volumes": [
    {
      "name": "application",
      "host": {
        "sourcePath": "/var/app/current/application"
      }
    }
  ],
  "containerDefinitions": [
    {
      "name": "php-fpm",
      "image": "123456.dkr.ecr.ap-northeast-1.amazonaws.com/php-fpm:latest",
      "essential": true,
      "memory": 128,
      "mountPoints": [
        {
          "sourceVolume": "application",
          "containerPath": "/var/www"
        }
      ]
    },
    {
      "name": "nginx",
      "image": "123456.dkr.ecr.ap-northeast-1.amazonaws.com/nginx:latest",
      "essential": true,
      "memory": 128,
      "mountPoints": [
        {
          "sourceVolume": "application",
          "containerPath": "/var/www"
        }
      ],
      "links": [
        "php-fpm"
      ],
      "portMappings": [
        {
          "hostPort": 80,
          "containerPort": 80
        }
      ]
    },
    {
      "name": "php-worker",
      "image": "123456.dkr.ecr.ap-northeast-1.amazonaws.com/php-worker:latest",
      "essential": true,
      "memory": 128,
      "mountPoints": [
        {
          "sourceVolume": "application",
          "containerPath": "/var/www"
        }
      ],
      "links": [
        "php-fpm"
      ]
    }
  ]
}
```

After deployment, the elastic beanstalk health is OK and there was nothing error, but nothing is displayed on the site.
Do I need to work on anything else?

**Comment by comlaterra** [comment]
Now that there is an `aws` service, I've followed what the documentaiton says: http://laradock.io/documentation/#use-aws also what AWS documentaiton explains, looks is up and running but:
I get Forbidden on the root url of the project
and a double laravel error without any log in any other path.

Any help is welcome, a link, some pointer og where to research to...?

**Comment by chkm8** [comment]
The aws service was meant to provide easy tools to access aws using eb ssh. Are you having Forbidden error when you access your environment url? I also had trouble in setting it up first but I forgot the specific errors but what I learned was that make sure that you setup the security group of each containers, they should have access with port 80 and 443. Also you need to enable that the listener to your eb instance both 80 and 443.

**Comment by comlaterra** [comment]
Thanks @chkm8 !
Well, I've changed the aws config to access to teh public folder, and I can see the laraval errors... without log (?!) and with two headers (?)
![screen shot 2017-09-21 at 11 46 19](https://user-images.githubusercontent.com/1207987/30691024-91414a26-9ec6-11e7-8b3f-226d3ebf8fa4.png)

I'm also trying to access to the logs, but I can't.
Any idea/points where I can search self-educate to fix this issue?

Thanks!

**Comment by chkm8** [comment]
How did you deploy your project to elastic beanstalk? Can you do tail -f log <apache2 docker instance inside eb server>?

**Comment by comlaterra** [comment]
I've deployed completely with eb cli. Followed the instructions:
`docker-compose up -d aws` then `docker-compose exec aws bash` and the rest from `eb cli`.
So `eb init` and `eb deploy --staged`. 
Do I need to have a `Dockerrun.aws.json` file? I though the point is that the cli makes up actions from the docker-compose.
`eb logs | grep apache2` doesn't returns anything, the logs in general looks good.

Thanks! I keep researching!

**Comment by chkm8** [comment]
If you wanted to deply your application using laradock, yes you need the Dockerrun.aws.json, there is a sample for that in the comment above on this thread. So if you dont use Dockerrun.aws.json, I assumed that when you did eb init you choose php as environment right? I havent tried deploying it in that way because I use the multiple docker options and I provided the Dockerrun.aws.json.

**Comment by codeasashu** [comment]
I am using nginx, php-fpm, mariadb for one of the project. I followed the steps suggested by @davehensley and can see a successful deployment. (I am using multi docker instance in aws ebs)

However, I was unable to see any nginx config when I `exec`'d into it. I see nginx.conf but no conf.d/default.conf. I think I have to write a Dockerfile, using `ADD` command to manually copy the config to `/etc/nginx/conf.d`. I will soon share the Dockerfile

**Comment by natar10** [comment]
@canfone Hey! Im having the same problem with deploy being on /var/app/current can you point me or give me some tips to solve this? thanks
